R/flags.R

```diff
@@ -57,3 +57,114 @@ add_flags <- function(x) {
   }
   x
 }
+
+#' Extract flags from yspec meta data
+#' 
+#' Use `ys_flags_chr()` to get a character vector of all columns returned by 
+#' `ys_flags()`.
+#' 
+#' @param x a yspec object. 
+#' @param ... tidy-select specification of flag names to select.
+#' 
+#' @return 
+#' `ys_flags()` returns a named list of flag column names. `ys_flags_chr()`
+#' returns a character vector of all flag columns returned by `ys_flags()`.
+#' 
+#' @examples
+#' 
+#' spec <- ys_help$spec()
+#' 
+#' ys_flags(spec, nm)
+#' 
+#' ys_flags_chr(spec, nm) 
+#' 
+#' @seealso [ys_select_fl()], [ys_factors_fl()]
+#' 
+#' @md
+#' @export
+ys_flags <- function(x, ...) {
+  flags <- pull_meta(x, "flags") 
+  if (!length(match.call(expand.dots = FALSE)$...)) {
+    return(flags)
+  }
+  pos <- eval_select(expr(c(...)), flags)
+  res <- flags[pos]
+  names(res) <- names(pos)
+  res
+}
+
+#' @md
+#' @rdname ys_flags
+#' @export
+ys_flags_chr <- function(x, ...) {
+  flags <- ys_flags(x, ...)
+  unlist(flags, use.names = FALSE)
+}
+
+#' Select items from a yspec object via flags
+#' 
+#' @inheritParams ys_flags_chr
+#' 
+#' @return 
+#' A yspec object containing columns named in passed through `...`. 
+#' 
+#' @examples
+#' 
+#' spec <- ys_help$spec()
+#' 
+#' ys_select_fl(spec, nm, times)
+#' 
+#' @seealso [ys_flags()], [ys_flags_chr()]
+#' 
+#' @md
+#' @export
+ys_select_fl <- function(x, ...) {
+  what <- ys_flags_chr(x, ...)
+  ys_select(x, tidyselect::all_of(what))
+}
+
+
+#' Add factors based on flags
+#' 
+#' The user passes a data set to modify, a `yspec` object, and 
+#' the names of flags in the yspec object to create factors
+#' in the data set using [ys_flags()].
+#' 
+#' @inheritParams ys_factors
+#' @param ... tidy-select specification of flag names to select.
+#' 
+#' @details
+#' Only eligible columns will be considered for factor creation (either 
+#' a `values` field is present or the `make_factor` flag is set). 
+#' Therefore, a flag containing a general list of data columns 
+#' (potentially continuous and discrete) could be passed without 
+#' generating an error. 
+#' 
+#' @examples
+#' data <- ys_help$data()
+#' spec <- ys_help$spec()
+#' 
+#' names(ys_flags(spec))
+#' 
+#' data <- ys_factors_fl(data, spec, cat)
+#' 
+#' head(data, 2)
+#' 
+#' data <- ys_factors_fl(data, spec, covariate)
+#' 
+#' head(data, 2)
+#' 
+#' @seealso [ys_factors()], [ys_add_factors()]
+#' @md
+#' @export
+ys_factors_fl <- function(data, spec, ..., .strict = TRUE) {
+  flags <- ys_flags_chr(spec, ...)
+  fct_ok <- map_lgl(spec, ~ isTRUE(.x[["make_factor"]]))
+  has_values <- map_lgl(spec, ~!is.null(.x[["values"]]))
+  eligible <- names(which(has_values | fct_ok))
```

Comment:
Same logic that we use in the factors functions; it wasn't quite flexible enough to just make this happen; but shouldn't be too bad to maintain.

Reply:
Trying to be permissive here to avoid need to have a very specific flag setup.